### PR TITLE
Stagger derivatives fixes

### PR DESCRIPTION
**Pull request by dschwoerer**
I thought I had written and upstreamed an MMS derivatives_flux test - but now I cannot find it.

Without checks, this results in potentially wrong results, with checks this throws.

### CODE REVIEW

**Comment by dschwoerer** [comment]
I found the test, it is in [ds_next](https://github.com/boutproject/BOUT-dev/tree/ds_next/tests/MMS/derivatives_flux) - but it fails with a very strange error:
```
Traceback (most recent call last):
  File "./runtest", line 104, in <module>
    runtests(functions,derivatives,directions,stag=False,msg="FDD")
  File "./runtest", line 24, in runtests
    errorlist+=runtest(nzs,guards,diff_func,diff,order,direction,fac,fs,vs,anas,locs,msg)
  File "./runtest", line 41, in runtest
    sim=diff_func(v,f,method=diff,outloc=locs[1])
  File "boutcore.pyx", line 1170, in boutcore.FDDX
RuntimeError: ====== Exception thrown ======
Unhandled staggering
====== Back trace ======
 -> Mesh::indexFDDX(Field3D, Field3D) on line 2651 of 'index_derivs.cxx'
====== Exception path ======
[bt] #1 pylib/boutcore.cpython-36m-x86_64-linux-gnu.so(_ZN13BoutExceptionC2EPKcz+0x29f) [0x7f9768647c9f]
BoutException::BoutException(char const*, ...) at /home/dave/soft/BOUT-dev/debug/src/sys/boutexception.cxx:134
[bt] #2 pylib/boutcore.cpython-36m-x86_64-linux-gnu.so(_ZN4Mesh9indexFDDXERK7Field3DS2_8CELL_LOC11DIFF_METHOD6REGION+0x22b7) [0x7f9768675787]
Mesh::indexFDDX(Field3D const&, Field3D const&, CELL_LOC, DIFF_METHOD, REGION) at /home/dave/soft/BOUT-dev/debug/src/mesh/index_derivs.cxx:2745
[bt] #3 pylib/boutcore.cpython-36m-x86_64-linux-gnu.so(_Z4FDDXRK7Field3DS1_8CELL_LOC11DIFF_METHOD6REGION+0x87) [0x7f97686419f7]
FDDX(Field3D const&, Field3D const&, CELL_LOC, DIFF_METHOD, REGION) at /home/dave/soft/BOUT-dev/debug/src/sys/derivs.cxx:398
[bt] #4 pylib/boutcore.cpython-36m-x86_64-linux-gnu.so(+0xec59f) [0x7f97685a659f]
__pyx_pf_8boutcore_12FDDX(_object*, __pyx_obj_8boutcore_Field3D*, __pyx_obj_8boutcore_Field3D*, _object*, _object*, _object*) [clone .isra.257] at /home/dave/soft/BOUT-dev/debug/tools/pylib/_boutcore_build/boutcore.cpp:19467
[bt] #5 pylib/boutcore.cpython-36m-x86_64-linux-gnu.so(+0xeca09) [0x7f97685a6a09]
__pyx_pw_8boutcore_13FDDX(_object*, _object*, _object*) at /home/dave/soft/BOUT-dev/debug/tools/pylib/_boutcore_build/boutcore.cpp:19388
[bt] #6 /lib64/libpython3.6m.so.1.0(_PyObject_FastCallDict+0x8b) [0x7f97705e381b]
_PyObject_FastCallDict at ??:?
[bt] #7 /lib64/libpython3.6m.so.1.0(_PyObject_FastCallKeywords+0x88) [0x7f9770651d98]
_PyObject_FastCallKeywords at ??:?
[bt] #8 /lib64/libpython3.6m.so.1.0(+0x182d41) [0x7f9770660d41]
_PyObject_CallMethod_SizeT at ??:?
[bt] #9 /lib64/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x12ae) [0x7f977069208e]
_PyEval_EvalFrameDefault at ??:?
[bt] #10 /lib64/libpython3.6m.so.1.0(+0x13ad5a) [0x7f9770618d5a]
PyFrame_New at ??:?
[bt] #11 /lib64/libpython3.6m.so.1.0(+0x182cc2) [0x7f9770660cc2]
_PyObject_CallMethod_SizeT at ??:?
[bt] #12 /lib64/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x34a) [0x7f977069112a]
_PyEval_EvalFrameDefault at ??:?
[bt] #13 /lib64/libpython3.6m.so.1.0(+0x10402b) [0x7f97705e202b]
_PyType_Lookup at ??:?
[bt] #14 /lib64/libpython3.6m.so.1.0(+0x13af10) [0x7f9770618f10]
PyFrame_New at ??:?
[bt] #15 /lib64/libpython3.6m.so.1.0(+0x182cc2) [0x7f9770660cc2]
_PyObject_CallMethod_SizeT at ??:?
[bt] #16 /lib64/libpython3.6m.so.1.0(_PyEval_EvalFrameDefault+0x12ae) [0x7f977069208e]
_PyEval_EvalFrameDefault at ??:?
[bt] #17 /lib64/libpython3.6m.so.1.0(PyEval_EvalCodeEx+0x318) [0x7f9770663058]
PyEval_EvalCodeEx at ??:?
[bt] #18 /lib64/libpython3.6m.so.1.0(PyEval_EvalCode+0x1b) [0x7f9770663f0b]
PyEval_EvalCode at ??:?
[bt] #19 /lib64/libpython3.6m.so.1.0(+0x21f7f2) [0x7f97706fd7f2]
PyErr_WarnExplicitObject at ??:?
[bt] #20 /lib64/libpython3.6m.so.1.0(PyRun_FileExFlags+0x9a) [0x7f97706ff3ca]
PyRun_FileExFlags at ??:?
[bt] #21 /lib64/libpython3.6m.so.1.0(PyRun_SimpleFileExFlags+0x1a8) [0x7f9770700128]
PyRun_SimpleFileExFlags at ??:?
[bt] #22 /lib64/libpython3.6m.so.1.0(Py_Main+0x6ce) [0x7f97707011fe]
Py_Main at ??:?
[bt] #23 python3(main+0x1f8) [0x564247251ca8]
[bt] #24 /lib64/libc.so.6(__libc_start_main+0xeb) [0x7f976f78811b]
__libc_start_main at ??:?
[bt] #25 python3(_start+0x2a) [0x564247251e1a]
```
Which is strange as all locations are `CELL_CENTER` ...

File: src/mesh/difops.cxx [hunk 1]
```diff
@@ -168,15 +168,15 @@ const Field3D Grad_parP(const Field3D &apar, const Field3D &f) {
 *******************************************************************************/
 
 const Field2D Vpar_Grad_par(const Field2D &v, const Field2D &f, const CELL_LOC outloc) {
-  return v.getCoordinates(outloc)->Vpar_Grad_par(v, f, outloc);
+  return f.getCoordinates(outloc)->Vpar_Grad_par(v, f, outloc);
```
Comment (d7919) [inline src/mesh/difops.cxx]: I think this change (and two others that are the same) only makes a difference if `f` and `v` are on different meshes, which we don't support. Is there a particular motivation for this change (e.g. to be consistent with other places in the code)?

Comment (dschwoerer) [reply src/mesh/difops.cxx]: Well, if they are staggered they are on the same mesh, but at different locations therein.

This worked before, and is used in e.g. STORM.

It also makes sense, as you want to advect quantities, then the advected quantity may ether be also staggered (e.g. velocity), or not (e.g. density).  In most cases, you would want the derivative at the advected quantity, which would be the one of `f`.

Sorry if I failed to address you comment, in which case I will try to clarify ...

Comment (d7919) [reply src/mesh/difops.cxx]: As we're passing `outloc` to `getCoordinates` in each case the location of the returned coordinates is entirely independent of which field we're using to fetch the coordinates from.

Comment (dschwoerer) [reply src/mesh/difops.cxx]: Ah I see.
`outloc` can be `CELL_DEFAULT` - in which case it matters.

Comment (d7919) [reply src/mesh/difops.cxx]: Ah yes, good point I forgot about `CELL_DEFAULT`.

**Pull request merged**